crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/metrics.rs

```diff
@@ -0,0 +1,140 @@
+use std::time::Instant;
+
+use super::*;
+
+/// Maximum amount of history we will track
+/// per remote node.
+const MAX_HISTORY: usize = 10;
+
+#[derive(Debug, Clone, Default)]
+/// Information about a remote node.
+struct Info {
+    /// Times we recorded errors for this node.
+    errors: VecDeque<Instant>,
+    /// Times we recorded initiates to this node.
+    initiates: VecDeque<Instant>,
+    /// Times we recorded remote rounds from this node.
+    remote_rounds: VecDeque<Instant>,
+    /// Times we recorded complete rounds for this node.
+    complete_rounds: VecDeque<Instant>,
+    /// Is this node currently in an active round?
+    current_round: bool,
+}
+
+#[derive(Debug, Default)]
+/// Metrics tracking for remote nodes to help
+/// choose which remote node to initiate the next round with.
+pub(super) struct Metrics {
+    /// Map of remote nodes.
+    map: HashMap<StateKey, Info>,
+    // Number of times we need to force initiate
+    // the next round.
+    force_initiates: u8,
+}
+
+/// Outcome of a gossip round.
+pub(super) enum Outcome {
+    Success(Instant),
+    Error(Instant),
+}
+
+impl Metrics {
+    #[cfg(test)]
+    pub(super) fn new() -> Self {
+        Self::default()
+    }
+
+    /// Record a gossip round has been initiated by us.
+    pub(super) fn record_initiate(&mut self, key: StateKey) {
+        let info = self.map.entry(key).or_default();
+        info.initiates.push_back(Instant::now());
+        if info.initiates.len() > MAX_HISTORY {
+            info.initiates.pop_front();
+        }
```

Comment:
Since VecDeque is implemented with a ring buffer it's annoying you can't just use it as such... I wonder if it'd be worth writing a minimal wrapper struct. Also, I think it might be more efficient to do the remove before the add if len = MAX_HISTORY.